corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+other:
+
+1. Merge and deploy a PR that configures multiplexing on an index.
+2. Execute an idempotent management command that updates the secondary index
+   from its primary counterpart.
+3. Merge and deploy a PR that disables multiplexing for the index, (now using
+   only the new index).
+4. Execute a management command to delete the old index.
+
+**Note**: the above steps are not limited to a single index at a time. That is,
+the implementation does not prohibit configuring multiplexing and reindexing
+multiple indexes at once.
+
+This reindex procedure is inherently safe because:
+
+- At any point in the process, the rollback procedure is a simple code change
+  (i.e. revert PR, deploy).
+- The operation responsible for populating the secondary index is idempotent
+  *and* decoupled from the index configuration, allowing it to undergo change
+  iterations without aborting the entire process (thereby losing reindex
+  progress).
+- Instructions for third party hosters can follow the same process that Dimagi
+  uses, which guarantees that any possible problems encountered by a third party
+  hoster are not outside the Dimagi main track.
+
+
+.. TODO: Future adapter documentation
+..
+.. - adapter patterns
+.. - querying with adapters
+.. - writing tests (use ``es_test``)
+
+----
+
+Elastic Client Adapters
+-----------------------
+
+The ``corehq.apps.es.client`` module encapsulates the CommCare HQ Elasticsearch
+client adapters. It implements a high-level Elasticsearch client protocol
+necessary to accomplish all interactions with the backend Elasticsearch cluster.
+Client adapters are split into two usage patterns, the "Management Adapter" and
+"Document Adapters".
+
+.. toctree::
+
+    Management Adapter
+    Document Adapters
+    Code Documentation
+
+
+Management Adapter
+''''''''''''''''''
+
+There is only one management adapter, ``ElasticManageAdapter``. This adapter is
+used for performing all cluster management tasks such as creating and updating
+indices and their mappings, changing index settings, changing cluster settings,
+etc.  This functionality is split into a separate class for a few reasons:
+
+1. The management adapter is responsible for low-level Elastic operations which
+   document adapters should never be performing because the scope of a document
+   adapter does not extend beyond a single index.
+2. Elasticsearch 5+ implements security features which limit the kinds of
+   operations a connection can be used for. The separation in these client
+   adapter classes is designed to fit into that model.
+
+The management adapter does not need any special parameters to work with, and
+can be instantiated and used directly:
+
+.. code-block:: python
+
+    adapter = ElasticManageAdapter()
+    adapter.index_create("books")
+    mapping = {"properties": {
+        "author": {"type": "text"},
+        "title": {"type": "text"},
+        "published": {"type": "date"},
+    }}
```

Comment:
How does this work if the old and new indexes have different mappings? Or does the management adaptor not support multiplexing? I imagine it would make sense for the management adapter to be tied to a single index otherwise it would make it hard to management them independently.

Reply:
Index creation and manipulation for multiplexing is not yet implemented, but it will certainly be done in a "multiplexing-aware" way at that time.